### ISSUE TITLE
For native commands, variables are not expanded if argument begins with `-` (`-Dkey=$var`)

### ISSUE THREAD

**Issue by MatejKafka** [PowerShell/PowerShell#21460]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

This issue follows up on one aspect of #6467. Many native commands use `-` prefix for named parameters, but either separate the parameter value with `=` (`-param=value`) or use no separator at all (e.g. GCC's `-Ddefine`).

Due to PowerShell parsing rules, even for native commands, anything starting with `-` is parsed as a possible parameter name, and PowerShell **does not expand variables in suspected parameter names**:
```powershell
> $var = "value"
> .\GetCommandLine.exe -key=$var -D$var
"D:\test\GetCommandLine.exe" -key=$var -D$var # no expansion of $var done
```

This behavior surprises me each time I encounter it, especially because it often does not result in any error, just the command silently doing something wrong:
```powershell
# WSL bash swallows the variable, and what is actually silently invoked is `wsl -- cmake -DCMAKE_BUILD_TYPE=`
wsl -- cmake -DCMAKE_BUILD_TYPE=$BuildType
```

I would argue that it would be much more intuitive to disable special parsing of arguments for native executables and treat tokens starting with `-` as an ordinary argument – native commands don't have a consistent way of parsing arguments, and assuming anything about the way they're parsed (except for `argv` splitting rules) will imo always lead to hard-to-understand edge cases.

---

GetCommandLine is the following C program, which just prints the command line exactly as it receives it:
```c
#include <Windows.h>
#include <stdio.h>

int main() {
	wprintf(L"%ls\n", GetCommandLineW());
}
```

### Expected behavior

```console
Variables are expanded in arguments to native commands, even when the argument begins with `-`.
```

### Actual behavior

```console
No variable expansions is done.
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.5.0-preview.2
PSEdition                      Core
GitCommitId                    7.5.0-preview.2
OS                             Microsoft Windows 10.0.19045
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by 237dmitry** [comment]
`--key=$value` parameters work. I could not find application with single  `-key=$value` syntax

![ss](https://github.com/PowerShell/PowerShell/assets/78153320/faf5ae78-b6ed-4d7a-bf51-2f2015bf6858)

**Comment by MartinGC94** [comment]
If the language allows people to use variables in parameter names then people will start doing things like: `$A = "Action"; Get-ChildItem C:\ -Error$A Ignore -Warning$A Ignore` and I don't like the thought of having to deal with scripts written like that.  
If you are getting surprised when a variable isn't being expanded then I'd advise you to use the syntax highlighting to your advantage.

**Comment by MatejKafka** [comment]
@MartinGC94 No, I don't want this for built-in commands, where the parser unambiguously knows what is a parameter name and what is a parameter value. However, for native commands, parser cannot ever reliably know that, and it seems more intuitive to just give up and allow expanding variables everywhere instead of allowing variables in some parameter values but not others.

**Comment by MatejKafka** [comment]
@MartinGC94 Ad syntax highlighting, it does not help that everything except PSReadline (I tested VS Code, NeoVim and Sublime Text) highlights the variable incorrectly, as if it was expanded.

**Comment by MatejKafka** [comment]
Hmm, that's a good point, that making the parser depend on knowing if it's parsing a native command or a cmdlet is probably not feasible, and would be a big can of worms even if theoretically possible.

Nevertheless, I would still prefer if variables would be expanded everywhere, even in parameter names, in hope that no one in their right mind will actually exploit it in the way you've shown above. While I agree that it would be nice to disallow variables in parameter names, I consider it more important for variable expansion to behave consistently, following the principle of least astonishment – if users encounter `ls -Error$var Ignore`, I believe most will be able to correctly guess at the behavior, while the current behavior with respect to native commands (needing to quote parameters starting with `-`, but not parameters starting with `--` or `/`) is unintuitive even to experienced users and easy to overlook when encountered, resulting in subtly broken scripts.

**Comment by SteveL-MSFT** [comment]
The workaround is to use quotes around the variable:

```sh
/bin/echo -d"$var"
/bin/echo -d="$var"
```
